File: app/src/main/java/com/example/newbiechen/ireader/ui/activity/MainActivity.java
package com.example.newbiechen.ireader.ui.activity;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.Toast;

import com.example.newbiechen.ireader.R;
import com.example.newbiechen.ireader.ui.base.BaseActivity;
import com.example.newbiechen.ireader.utils.PermissionsChecker;
import com.example.newbiechen.ireader.utils.ToastUtils;

public class MainActivity extends BaseActivity {
    private static final int WAIT_INTERVAL = 2000;
    private static final int PERMISSIONS_REQUEST_STORAGE = 1;

    static final String[] PERMISSIONS = {
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
    };
    private PermissionsChecker mPermissionsChecker;
    private boolean isPrepareFinish = false;

    @Override
    protected int getContentId() {
        return R.layout.activity_main;
    }

    @Override
    protected void setUpToolbar(Toolbar toolbar) {
        super.setUpToolbar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(false);
        getSupportActionBar().setTitle("阅读器");
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_main, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        Class<?> activityCls = null;
        if (id == R.id.action_scan_local_book) {
            if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
                if (mPermissionsChecker == null) {
                    mPermissionsChecker = new PermissionsChecker(this);
                }
                //获取读取和写入SD卡的权限
                if (mPermissionsChecker.lacksPermissions(PERMISSIONS)) {
                    //请求权限
                    ActivityCompat.requestPermissions(this, PERMISSIONS, PERMISSIONS_REQUEST_STORAGE);
                    return super.onOptionsItemSelected(item);
                }
            }
            activityCls = FileSystemActivity.class;
        }
        if (activityCls != null) {
            Intent intent = new Intent(this, activityCls);
            startActivity(intent);
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == PERMISSIONS_REQUEST_STORAGE) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                //跳转到 FileSystemActivity
                Intent intent = new Intent(this, FileSystemActivity.class);
                startActivity(intent);

            } else {
                ToastUtils.show("用户拒绝开启读写权限");
            }
        }
    }

    @Override
    public void onBackPressed() {
        if (!isPrepareFinish) {
            new Handler().postDelayed(
                    () -> isPrepareFinish = false, WAIT_INTERVAL
            );
            isPrepareFinish = true;
            Toast.makeText(this, "再按一次退出", Toast.LENGTH_SHORT).show();
        } else {
            super.onBackPressed();
        }
    }
}

File: app/src/main/java/com/wangzhen/reader/ui/activity/MainActivity.kt
package com.wangzhen.reader.ui.activity

import android.Manifest
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.provider.Settings
import android.view.View
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.GridLayoutManager
import com.wangzhen.adapter.base.RecyclerItem
import com.wangzhen.permission.PermissionManager.request
import com.wangzhen.permission.callback.AbsPermissionCallback
import com.wangzhen.reader.R
import com.wangzhen.reader.base.BookRepository
import com.wangzhen.reader.databinding.ActivityMainBinding
import com.wangzhen.reader.databinding.HeaderBookShelfBinding
import com.wangzhen.reader.model.bean.CollBookBean
import com.wangzhen.reader.ui.adapter.BookShelfAdapter
import com.wangzhen.reader.ui.base.BaseActivity
import com.wangzhen.reader.utils.toast
import java.util.Locale

/**
 * MainActivity
 * Created by wangzhen on 2023/4/11
 */
class MainActivity : BaseActivity() {
    private lateinit var binding: ActivityMainBinding
    private lateinit var shelfAdapter: BookShelfAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        fitLightStatusBar()
        setUpAdapter()
        loadBooks()
    }

    private fun checkPermissions() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
            checkExternalPermissions()
        } else {
            if (!Environment.isExternalStorageManager()) {
                launcher.launch(Intent(Settings.ACTION_MANAGE_ALL_FILES_ACCESS_PERMISSION))
            } else {
                openFileSystem()
            }
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.R)
    private val launcher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
            if (Environment.isExternalStorageManager()) {
                openFileSystem()
            } else {
                "请授予所有文件访问权限".toast()
            }
        }

    private fun checkExternalPermissions() {
        request(this, object : AbsPermissionCallback() {
            override fun onGrant(permissions: Array<String>) {
                openFileSystem()
            }

            override fun onDeny(
                deniedPermissions: Array<String>, neverAskPermissions: Array<String>
            ) {
                "请授予存储访问权限".toast()
            }
        }, Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE)
    }

    private fun openFileSystem() {
        startActivity(Intent(this, FileSystemActivity::class.java))
    }

    private fun loadBooks() {
        shelfAdapter.setData(BookRepository.instance.collBooks)
    }

    private fun setUpAdapter() {
        with(binding.recycler) {
            layoutManager = GridLayoutManager(context, 3)
            adapter = BookShelfAdapter(BookRepository.instance.collBooks).apply {
                shelfAdapter = this
                setOnClickCallback { _, pos ->
                    ReadActivity.startActivity(this@MainActivity, datas[pos])
                }
                setOnLongClickCallback { _, pos ->
                    deleteBook(datas[pos])
                }
                addHeader(object : RecyclerItem() {
                    override fun layout() = R.layout.header_book_shelf
                    override fun onViewCreated(itemView: View) {
                        HeaderBookShelfBinding.bind(itemView).apply {
                            btnChooseFiles.setOnClickListener { checkPermissions() }
                            btnWifiTransfer.setOnClickListener {
                                startActivity(Intent(it.context, WifiTransferActivity::class.java))
                            }
                        }
                    }
                }.onCreateView(binding.root))
                setEmpty(object : RecyclerItem() {
                    override fun layout() = R.layout.layout_bookshelf_empty
                }.onCreateView(binding.root))
            }
        }
    }

    private fun deleteBook(book: CollBookBean) {
        AlertDialog.Builder(this)
            .setMessage(String.format(Locale.getDefault(), "确定删除%s吗", book.title))
            .setNegativeButton("取消", null).setPositiveButton("确定") { _, _ ->
                BookRepository.instance.deleteBook(book)
                Toast.makeText(this, book.title + "已删除", Toast.LENGTH_SHORT).show()
                setUpAdapter()
            }.create().show()
    }

    override fun onResume() {
        super.onResume()
        loadBooks()
    }
}